Add cache of layered classloaders for test and run tasks

eatkins opened the pull request:
### usage

```scala
Compile / classLoaderLayeringStrategy := ClassLoaderLayeringStrategy.Flat
Compile / classLoaderLayeringStrategy := ClassLoaderLayeringStrategy.ScalaInstance
Compile / classLoaderLayeringStrategy := ClassLoaderLayeringStrategy.RuntimeDependencies

Test / classLoaderLayeringStrategy := ClassLoaderLayeringStrategy.Flat
Test / classLoaderLayeringStrategy := ClassLoaderLayeringStrategy.ScalaInstance
Test / classLoaderLayeringStrategy := ClassLoaderLayeringStrategy.RuntimeDependencies
Test / classLoaderLayeringStrategy := ClassLoaderLayeringStrategy.ShareRuntimeDependenciesLayerWithTestDependencies
Test / classLoaderLayeringStrategy := ClassLoaderLayeringStrategy.TestDependencies
```

### original description

This commit adds layered classloaders to sbt so that we can cache the classloaders for the portion of the classpath that never changes. There are two benefits to doing this:
1) performance: by reusing classloaders for the dependencies, we do not have to reload all of the potentially large library jar classes. This classloading dominates the startup time of the test and run tasks.
2) memory/metaspace utilization: sbt can quickly run out of memory or meta space if any of the tasks potentially leak classloaders (usually by leaking a thread). Caching the classloaders that load most of the classes can help. Especially if the leak is in the library classes. 

Here are some gifs that show off the performance improvements:

In this example, I just illustrate how quickly the test task runs when invoked repeatedly, even though the test has to initialize an Akka actor system:
![Output sample](https://media.giphy.com/media/2wY1SFTo9mtHgBUQHS/giphy.gif)
Note that after I cleared the cache, it got slow again on the next iteration.

Here, I generated 5000 source files in the above project. In each build, zinc has to check all of the source files for changes, which can take a long time if io must be done on all 5000 files. In the following to gifs, I show one terminal triggering a continuous build and the build running in the second window. In the first gif, I use no caching:
![Output sample](https://media.giphy.com/media/xksH0LoGCcfXtTLp1a/giphy.gif)

Here I use full caching:
![Output sample](https://media.giphy.com/media/2iqbwaHPW9I1OntxVY/giphy.gif)

eatkins committed:
Add LRUCache to sbt.internal

I am going to add a classloader cache to improve the startup performance
of the run and test tasks. To prevent the classloader cache from having
unbounded size, I'm adding a simple LRUCache implementation to sbt. An
important characteristic of the implementation of the cache is that when
entries are evicted, we run a callback to cleanup the entry. This allows
us to automatically cleanup any resources created by the entry.

This is a pretty naive implementation that uses an array of entries that
it manipulates as elements are removed/accessed. In general, I expect
these caches to be quite small <= 4 elements, so the storage overhead /
performance of the simple implementation should be quite good. If
performance ever becomes an issue, we can specialzed LRUCache.apply to
use a different implementation for caches with large limits.

eatkins committed:
Add ClassLoaderCache

In order to speed up the start up time of the test and run tasks, I'm
introducing a ClassLoaderCache that can be used to avoid reloading the
classes in the project dependencies (which includes the scala library).
I made the api as minimal as possible so that we can iterate on the
implementation without breaking binary compatibility. This feature will
be gated on a feature flag, so I'm not concerned with the cache class
loaders being useable in every user configuration. Over time, I hope
that the CachedClassLoaders will be a drop in replacement for the
existing one-off class loaders*.

The LayeredClassLoader was adapted from the NativeCopyLoader. The main
difference is that the NativeCopyLoader extracts named shared libraries
into the task temp directory to ensure that the ephemeral libraries are
deleted after each task run. This is a problem if we are caching the
ClassLoader so for LayeredClassLoader I track the native libraries that
are extracted by the loader and I delete them either when the loader is
explicitly closed or in a shutdown hook.

* This of course means that we both must layer the class loaders
appropriately so that the project code is in a layer above the cached
loaders and we must correctly invalidate the cache when the project, or
its dependencies are updated.

eatkins committed:
Add Repository interface

I am going to be introducing multiple caches throughout sbt and I am
going to build these features out using this simple Repository
interface. The idea is that we access data by some key through the
Repository. This allows us to use the strategy pattern to easily switch
the runtime implementation of how to get the data.

eatkins committed:
Add scripted tests for classloader-cache

Normally I'd include these with the previous commit, but the diff is so
large that I put them in their own commit. The tests handle 5 scenarios:

1) akka-actor-system -- a project that has Akka as a dependency and a
   simple main method that creates and terminates an ActorSystem. What
   is interesting about this test is that if scriptedBufferLog := false,
   we notice that the first call to run is slow, but subsequent calls to
   run and test are fast. The test does at least ensure that recycling
   the runtime layer in test works ok.

2) jni -- verifies that a project with native libraries will be able to
   load the library with each run. It actually swaps out the underlying
   library so that the it really ensures that the library is reloaded
   between runs.

3) library-mismatch -- verifies that the layered classloaders can work
   when the test dependencies are incompatible with the runtime
   dependencies. In this test, the test dependencies use an api in a
   library called foo-lib that isn't available in the version used by
   the runtime dependencies. Because of this incompatibility, the test
   will not work if Test / layeringStrategy := LayeringStrategy.Full.

4) scalatest -- verifies that a test runs using the scalatest framework
5) utest -- verifies that a test runs using the utest framework

The reason for (4) and (5) is to ensure that both the in sourced test
frameworks and external frameworks work with the new loaders.

eatkins committed:
Add TaskRepository to manage ClassLoaderCache

We want the user to be able to invalidate the classloader cache in the
event that it somehow gets in a bad state. The cache is, however,
defined in multiple configurations, so there are in fact many
ClassLoaderCache instances that are managed by sbt. To make this sane, I
add a global cache that is keyed by a TaskKey[_] and can return
arbitrary data back. Invalidating all of the ClassLoaderCache instances
is then as straightforward as just replacing the TaskRepository
instance.

I also went ahead and unified the management of the global file tree
repository. Instead of having to specifically clear the file tree
repository or the classloader cache, the user can now invalidate both
with the new clearCaches command.

eatkins committed:
Add LayeringStrategy

This introduces a new trait LayeringStrategy that is used to configure
how sbt constructs the ClassLoaders used by the run and test tasks. In
addition to defining the various options, I try to give a good high
level overview of the problem that the LayeringStrategy is intended to
address in its scaladoc.

eatkins committed:
Cleanup scripted tests

The sleeps in source-dependencies/export-jars are no longer necessary
and just slow the test down. I also fixed minor syntax issues in
scala-instance-classloader and fixed a thread leak in the test.

eatkins committed:
Improve TestFramework error reporting

It is possible for the test task to fail because of issues with the
layered ClassLoaders. When this occurs and is detectable, I try to
provide a useful error message that will help the user fix the issue.

eatkins committed:
Use layered ClassLoaders in run and test tasks

Using the data structures that I added in the previous commits, it is
now possible to rework the run and test task to use (configurable)
layered class loaders. The layering strategy is globally set to
LayeringStrategy.Default. The default strategy leads to what is
effectively a three layered ClassLoader for the both the test and run
tasks. The first layer contains the scala instance (and test framework
loader in the test task). The second layer contains all of the
dependencies for the configuration while the third layer contains the
project artifacts.

The layering strategy is very easily changed both at the Global or
Configuration level, e.g. adding
Test / layeringStrategy := LayeringStrategy.Flat
to the project build.sbt will make the test task not even use the scala
instance and instead a create a single layer containing the full
classpath of the test task.

I also tried to ensure that all of the ClassLoaders have good toString
overrides so that it's easy to see how the ClassLoader is constructed
with, e.g. `show testLoader`, in the sbt console.

In this commit, the ClassLoaderCache instances are settings. In the next
commit, I make them tasks so that we can easily clear out the caches
with a command.

eed3si9n commented:
sbt currently depends on ScalaCache and [caffeine](https://github.com/ben-manes/caffeine). Do you think we could use them for this purpose?

He-Pin commented:
Thanks, @eatkins really great! With it already depends on caffeine, so maybe reuse of them? other things are:
1. limit the max opening files
2. when evicted, close the file.

eatkins commented:
@eed3si9n we could certainly use caffeine to replace LRUCache. Beyond that, I'm not completely sure that we'd get much additional value out of the caching libraries. The actual caching logic is minimal and I put everything behind interfaces that don't expose whether or not a cache is used to retrieve the result.

eatkins commented:
@kerr is there a specific issue that you've seen with sbt opening too many files? As far as I know, sbt shouldn't be holding handles to open files regardless of whether or not a cache is used. FWIW, all of the caches that I've added to sbt cleanup the cached resources on eviction or when the cache is explicitly closed.

eed3si9n commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -373,6 +375,7 @@ object Defaults extends BuildCommon {
         excludeFilter in unmanagedSources
       )
       .value,
+    layeringStrategy in ConfigGlobal := LayeringStrategy.Default,
```
This should go to global scope so we can do `ThisBuild / layeringStrategy`

eed3si9n commented on main/src/main/scala/sbt/LayeringStrategy.scala:
```diff
@@ -0,0 +1,115 @@
+ * In general, this should only happen if the user explicitly overrides the thread context
+ * ClassLoader or uses reflection to manipulate classes loaded by different loaders.
+ */
+sealed trait LayeringStrategy
```
Contraband?

eed3si9n commented on main/src/main/scala/sbt/Keys.scala:
```diff
@@ -264,6 +264,7 @@ object Keys {
   val bgRunMain = inputKey[JobHandle]("Start a provided main class as a background job")
   val fgRunMain = inputKey[Unit]("Start a provided main class as a foreground job")
   val bgCopyClasspath = settingKey[Boolean]("Copies classpath on bgRun to prevent conflict.")
+  val layeringStrategy = settingKey[LayeringStrategy]("Creates the classloader layering strategy for the particular configuration.")
```
Maybe we should come up with more verbose or intuitive naming for this key and class?

eatkins replied:
I changed it to [cC]lassLoaderLayeringStrategy which is definitely verbose and hopefully more intuitive.

eed3si9n commented on main/src/main/scala/sbt/LayeringStrategy.scala:
```diff
@@ -0,0 +1,115 @@
+ * Provides instances of [[LayeringStrategy]] that can be used to define the ClassLoader used by
+ * [[Keys.run]], [[Keys.test]] or any other task that runs java code inside of the sbt jvm.
+ */
+object LayeringStrategy {
```
Also let's think more about the enum names.

eatkins replied:
https://github.com/sbt/sbt/pull/4476/commits/25414683740c4e57e53fec31787dc0986fafa785

eed3si9n commented:
It would be interesting to see how slf4j / logback behaves since they use classpath as configuration.

eatkins commented:
<s>I can't reproduce the travis failures either on my mac or on my linux vm using adopt jdk or just regular openjdk. It is really weird.
I fixed the travis failures by removing `sudo: false` and `dist: trusty` from the .travis.yml. Travis is removing `sudo: false` anyway. I'm not sure why we were specifying the dist key.</s>

Actually there was a bug in initializing the file tree repository that caused the travis failures.

eatkins committed:
Fix FileRepository implementation for scripted

I was seeing spurious travis failures and I finally tracked it down to
the fact that in some cases, the project metabuild would sometimes use a
caching file tree repository instead of a polling repository. This
caused problems because the caching repository can take a few milliseconds
to detect changes in a directory. Because scripted copies the project
sources to the temporary test directory, it was possible for the project
meta build compilation to be initiated before the cache was aware of all
of the files.

The reason this happened was because scripted would create a state where
the remaining commands looked like:
List(sbtPopOnFailure, resumeFromFailure, notifyUsersAboutShell, iflast shell, ~compile, < 41684)
The ~compile command was causing the continuous flag to get set to true
which caused the default file tree repository task to return the caching
version. The reason for the continuous flag was so that when sbt is
started in a non-interactive mode where the command is to be repeated,
then we use the caching file tree repository. To support this use case,
we just need to check that the last command begins with `~`, not that
_any_ command begins with `~`.

jvican commented:
Nice to see this change implemented here @eatkins A layered classloader infrastructure to run tests is something it's been on my todo list for a long time, I will probably experiment and merge it in bloop soon too. It's nice to see this kind of experimentation in the sbt main repository.

eatkins commented:
@jvican thanks for the feedback. This work did start as a plugin, but I mentioned to @eed3si9n that I was working on the plugin and he mentioned that adding this may fix or at least mitigate some longstanding performance/memory issues in sbt so I went ahead and directly embedded the plugin into sbt.

Long term, it may make sense to move some of this code into the util project so that it can be used by other projects. Right now, it is too much of a pain to work across repos which is why I implemented everything directly in sbt itself.

eatkins committed:
JDK 11 fixes

In JDK 11, String.lines returns java.util.stream.Stream[String]. Switch
to linesIterator instead.

eatkins committed:
Rename some layering related fields/classes

In code review, @eed3si9n suggested that I switch to a more verbose and
descriptive naming scheme. In addition to trying to make layers more
descriptive, I also made the various layer case objects extend the
relevant layers so it's more clear what the layer should look like.

eatkins commented:
@eed3si9n I think I addressed your review feedback and I feel comfortable with this PR now.

eed3si9n commented on main/src/main/scala/sbt/internal/ClassLoaders.scala:
```diff
@@ -0,0 +1,213 @@
+            (true, true)
+          case ClassLoaderLayeringStrategy.ScalaInstance       => (false, false)
+          case ClassLoaderLayeringStrategy.RuntimeDependencies => (true, false)
+          case _                                               => (false, true)
```
Does `Default` fall into this pattern, if so it might need a distinct name.

eatkins committed:
Set classLoaderLayeringStrategy in relevant configs

Previously, the ClassLoaderLayeringStrategy was set globally. This
didn't really make sense because the Runtime and Test configs had
different strategies available (Test being a superset of Runtime).
Instead, we now set the layering strategy in the Runtime and Test
configurations directly. In doing this, we can eliminate the Default
ClassLoaderLayeringStrategy. Previously this had existed so that we
could set the layering strategy globally and have it do the right thing
in both test and runtime.

To implement this, I factored out the logic for generating the layered
classloader in the test task and shared it with the runtime task. I did
this because I realized that Test / run is a thing. Previously I had
been operating under the assumption that the runner would never include
the test dependencies. Once I realized this, it made sense to combine
the logic in both tasks.

As a bonus, I only allow the layering strategies that explicitly make
sense to be set in each configuration. If the user sets an invalid
strategy, an error will be thrown that specifies the valid strategies
for the task.

I also added ScalaInstance as an option for the runtime layer. It was an
oversight that this was left out.

eatkins committed:
Bump classloader-cache test to scala 2.12.8

These were written before we upgraded scala in the develop branch. I'm
not doing this as a fixup to keep the PR clean.

eatkins commented:
The idea behind Default was that it could be set globally and then any task that needed a layered classloader could choose what to do when Default was specified. I realized that Default really only applied at the configuration level, so it made sense to just explicitly set the classLoaderLayeringStrategy in the Compile and Test configurations. This is done in 
~~https://github.com/sbt/sbt/pull/4476/commits/87a0c58790396e97200e241a188f28425b7dabf5~~ 
~~https://github.com/sbt/sbt/pull/4476/commits/2e0dbfc7b7016786ebb4cbd135cb823b764ac40d~~ 
https://github.com/sbt/sbt/commit/ab5f14c5f03d264ff7e13cb3b4f9a81bb6ce9a1d.

eed3si9n reviewed (dismissed):
I think we are good to go on this.
Thanks for the contribution!

eatkins commented:
@eed3si9n I just force pushed the branch rebased against origin/develop. I need another approval to actually merge once CI finishes.

eed3si9n commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -1764,24 +1760,44 @@ object Defaults extends BuildCommon {
+  lazy val runnerSettings: Seq[Setting[_]] = {
+    val unscoped: Seq[Def.Setting[_]] =
+      Seq(runner := ClassLoaders.runner.value, forkOptions := forkOptionsTask.value)
+    inConfig(Compile)(unscoped) ++ inConfig(Test)(unscoped)
```
`runnerSettings` is part of `configSettings`, so this shouldn't be scoped to a particular configuration.
This will break some of the plugins - https://github.com/sbt/sbt-validator/issues/35

eatkins replied:
Ok. I'll figure out how to fix this.

eatkins replied:
I was a little worried about this change actually.

retronym commented:
Let's warn people about the risks of leaking static state across `run` invocations.

For example, if `runMain my.App` executed `myLibrary.staticField = 42`, this would be still seen by subsequent `run` tasks.

The static state may retain large amounts of heap (for instance a cache with no timed eviction policy).

I would suggest that SBT's `ClassLoaderCache` itself should have a timed eviction policy to counteract these problems. We would probably like the classloader cache used for the compiler to have a long timeout (because JITting scalac takes 1-2 minutes!), and a smaller timeout for third-party library caches (because we have no idea how much memory they statically retain.)

eatkins commented:
@retronym there is a command, clearCaches, that flushes the classloader cache.

The third party library case is difficult to handle universally. Ideally, a library should provide handles to any resources whether they are caches or threads or whatever that allow them to be cleaned up after the test run completes. In the real world, however, I am sure many projects rely on third party libraries that use singletons and leak threads. The question partially becomes to what extent do we allow badly written third party libraries control the default behavior of sbt. My instinct is to implicitly encourage users to use libraries that don’t leak resources because we can make the sbt ux so much better in those cases. I do not, however, want users to grow overly frustrated with sbt if the defaults are not reliably working in the most important use cases. All of that is to say that we are still trying to figure out what the impact of changing the default layering strategy is. Things may change between now and 1.3.0 final, but we’re gathering some very valuable data in the meantime.

eatkins commented:
I should add that I do 100% agree that by sbt 1.3.0 final there should be a detailed guide to all the layering strategies: the use cases each supports and the drawbacks/issues that may be encountered.

Not to belabor the point, but thread leaks are a problem even if the cache is evicted since the thread references are kept alive. sbt has always had this problem.

The only true solution is fork :)
